File: src/python/multithreading/thread_local_storage.py
"""
Thread Local Storage (TLS) Example

This script demonstrates the use of threading.local() to create thread-specific
data storage. Each thread has its own independent copy of the data.

Key Concepts:
- threading.local() creates an object with thread-specific attributes
- Each thread sees only its own data, even with the same attribute names
- No synchronization needed - each thread has its own isolated storage

Use Cases:
- Database connections per thread
- User session data in web applications
- Request context in server applications
- Thread-specific configuration or state

Alternative Approaches:
- Global dictionary with thread IDs as keys (requires manual locking)
- Pass context explicitly through function arguments
- Use contextvars (Python 3.7+) for async-compatible context
"""

import threading

# Create a thread-local data object
thread_local_data = threading.local()


def worker(thread_id):
    """
    Worker function that demonstrates thread-local storage.

    Each thread:
    1. Stores its ID in thread-local storage
    2. Performs work that modifies the thread-local data
    3. Reads and displays its thread-local data
    """
    # Store thread-local data
    thread_local_data.value = thread_id

    # Simulate some work that modifies thread-local data
    some_work()

    # Read thread-local data (each thread sees only its own value)
    print(f"Thread {thread_id} has value: {thread_local_data.value}")


def some_work():
    """Modify the thread-local data."""
    thread_local_data.value *= 2


def main():
    num_threads = 5
    threads = [threading.Thread(target=worker, args=(i,)) for i in range(num_threads)]

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()


if __name__ == "__main__":
    main()

File: src/cpp/multithreading/thread_local_storage.cpp
/**
 * Thread Local Storage
 *
 * This example demonstrates thread_local storage - variables that have
 * a separate instance for each thread.
 *
 * Key concepts:
 * - thread_local variables have per-thread storage
 * - Each thread gets its own copy, initialized independently
 * - No synchronization needed - each thread owns its copy
 * - Useful for caching, error state, or thread-specific context
 *
 * Without thread_local, these variables would need mutex protection.
 */

#include <iostream>
#include <thread>
#include <vector>

// Each thread gets its own copy of this variable
thread_local int thread_data = 0;

void do_work() {
    thread_data *= 2;  // Safe - only this thread accesses this copy
}

void worker(int id) {
    // Initialize this thread's copy
    thread_data = id;

    // Perform work that modifies thread-local data
    do_work();

    // Each thread sees its own value
    std::cout << "Thread " << id << " has value: " << thread_data << "\n";
}

int main() {
    constexpr int num_threads = 5;
    std::vector<std::thread> threads;
    threads.reserve(num_threads);

    for (int i = 1; i <= num_threads; ++i) {
        threads.emplace_back(worker, i);
    }

    for (auto& t : threads) {
        t.join();
    }

    // Each thread sees: thread_id * 2
    // Thread 1 -> 2, Thread 2 -> 4, Thread 3 -> 6, etc.
    std::cout << "All threads complete.\n";

    return 0;
}
